> **Tip**: Welcome to the Investigate a Dataset project! You will find tips in quoted sections like this to help organize your approach to your investigation. Before submitting your project, it will be a good idea to go back through your report and remove these sections to make the presentation of your work as tidy as possible. First things first, you might want to double-click this Markdown cell and change the title so that it reflects your dataset and investigation.

# Project: Investigating a movie database

## Table of Contents
<ul>
<li><a href="#intro">Introduction</a></li>
<li><a href="#wrangling">Data Wrangling</a></li>
<li><a href="#eda">Exploratory Data Analysis</a></li>
<li><a href="#conclusions">Conclusions</a></li>
</ul>

<a id='intro'></a>
## Introduction

> **Tip**: In this section of the report, provide a brief introduction to the dataset you've selected for analysis. At the end of this section, describe the questions that you plan on exploring over the course of the report. Try to build your report around the analysis of at least one dependent variable and three independent variables. If you're not sure what questions to ask, then make sure you familiarize yourself with the dataset, its variables and the dataset context for ideas of what to explore.

> If you haven't yet selected and downloaded your data, make sure you do that first before coming back here. In order to work with the data in this workspace, you also need to upload it to the workspace. To do so, click on the jupyter icon in the upper left to be taken back to the workspace directory. There should be an 'Upload' button in the upper right that will let you add your data file(s) to the workspace. You can then click on the .ipynb file name to come back here.

> Question 1: Which genres are most popular from year to year?
> <br>Question 2: What is the distribution of movies making a profit over their budget?
> <br>Question 3: Does a higher budget increase the popularity and vote of a movie?

In [26]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline

<a id='wrangling'></a>
## Data Wrangling


### General Properties

> Let's load the data first and visualize the first few line:

In [27]:
df = pd.read_csv('tmdb-movies.csv')
df.head()

id    imdb_id  popularity     budget     revenue  \
0  135397  tt0369610   32.985763  150000000  1513528810   
1   76341  tt1392190   28.419936  150000000   378436354   
2  262500  tt2908446   13.112507  110000000   295238201   
3  140607  tt2488496   11.173104  200000000  2068178225   
4  168259  tt2820852    9.335014  190000000  1506249360   

                 original_title  \
0                Jurassic World   
1            Mad Max: Fury Road   
2                     Insurgent   
3  Star Wars: The Force Awakens   
4                     Furious 7   

                                                cast  \
0  Chris Pratt|Bryce Dallas Howard|Irrfan Khan|Vi...   
1  Tom Hardy|Charlize Theron|Hugh Keays-Byrne|Nic...   
2  Shailene Woodley|Theo James|Kate Winslet|Ansel...   
3  Harrison Ford|Mark Hamill|Carrie Fisher|Adam D...   
4  Vin Diesel|Paul Walker|Jason Statham|Michelle ...   

                                            homepage          director  \
0                      http://www.jurassicworld.com/   Colin Trevorrow   
1                        http://www.madmaxmovie.com/     George Miller   
2     http://www.thedivergentseries.movie/#insurgent  Robert Schwentke   
3  http://www.starwars.com/films/star-wars-episod...       J.J. Abrams   
4                           http://www.furious7.com/         James Wan   

                         tagline      ...       \
0              The park is open.      ...        
1             What a Lovely Day.      ...        
2     One Choice Can Destroy You      ...        
3  Every generation has a story.      ...        
4            Vengeance Hits Home      ...        

                                            overview runtime  \
0  Twenty-two years after the events of Jurassic ...     124   
1  An apocalyptic story set in the furthest reach...     120   
2  Beatrice Prior must confront her inner demons ...     119   
3  Thirty years after defeating the Galactic Empi...     136   
4  Deckard Shaw seeks revenge against Dominic Tor...     137   

                                      genres  \
0  Action|Adventure|Science Fiction|Thriller   
1  Action|Adventure|Science Fiction|Thriller   
2         Adventure|Science Fiction|Thriller   
3   Action|Adventure|Science Fiction|Fantasy   
4                      Action|Crime|Thriller   

                                production_companies release_date vote_count  \
0  Universal Studios|Amblin Entertainment|Legenda...       6/9/15       5562   
1  Village Roadshow Pictures|Kennedy Miller Produ...      5/13/15       6185   
2  Summit Entertainment|Mandeville Films|Red Wago...      3/18/15       2480   
3          Lucasfilm|Truenorth Productions|Bad Robot     12/15/15       5292   
4  Universal Pictures|Original Film|Media Rights ...       4/1/15       2947   

   vote_average  release_year    budget_adj   revenue_adj  
0           6.5          2015  1.379999e+08  1.392446e+09  
1           7.1          2015  1.379999e+08  3.481613e+08  
2           6.3          2015  1.012000e+08  2.716190e+08  
3           7.5          2015  1.839999e+08  1.902723e+09  
4           7.3          2015  1.747999e+08  1.385749e+09  

[5 rows x 21 columns]

>Our first observation is that the 'genres' column will need to be stripped out as multiple genres are consolidated in one line.
<br>Next we will look for instances of missing or possibly errant data.

In [28]:
df.shape

(10866, 21)

> There are 10866 rows in this file and 21 columns.
<br>Let's analyse whether there are any missing data that need to be removed in order to answer the three questions for our project:

In [29]:
df.isnull().sum()

id                         0
imdb_id                   10
popularity                 0
budget                     0
revenue                    0
original_title             0
cast                      76
homepage                7930
director                  44
tagline                 2824
keywords                1493
overview                   4
runtime                    0
genres                    23
production_companies    1030
release_date               0
vote_count                 0
vote_average               0
release_year               0
budget_adj                 0
revenue_adj                0
dtype: int64

> There are 23 rows for genres that are missing which should be removed to answer the first questions on our project.
<br> We should now identify whether there are any duplicates rows to remove them:

In [30]:
df[df.duplicated()==True]

id    imdb_id  popularity    budget  revenue original_title  \
2090  42194  tt0411951     0.59643  30000000   967000         TEKKEN   

                                                   cast homepage  \
2090  Jon Foo|Kelly Overton|Cary-Hiroyuki Tagawa|Ian...      NaN   

              director              tagline     ...       \
2090  Dwight H. Little  Survival is no game     ...        

                                               overview runtime  \
2090  In the year of 2039, after World Wars destroy ...      92   

                                           genres    production_companies  \
2090  Crime|Drama|Action|Thriller|Science Fiction  Namco|Light Song Films   

     release_date vote_count  vote_average  release_year  budget_adj  \
2090      3/20/10        110           5.0          2010  30000000.0   

      revenue_adj  
2090     967000.0  

[1 rows x 21 columns]

> One row appears to be duplicated in index 2090.
<br>Next we will look at the data types of the variables that will be used in our analysis:

In [31]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10866 entries, 0 to 10865
Data columns (total 21 columns):
id                      10866 non-null int64
imdb_id                 10856 non-null object
popularity              10866 non-null float64
budget                  10866 non-null int64
revenue                 10866 non-null int64
original_title          10866 non-null object
cast                    10790 non-null object
homepage                2936 non-null object
director                10822 non-null object
tagline                 8042 non-null object
keywords                9373 non-null object
overview                10862 non-null object
runtime                 10866 non-null int64
genres                  10843 non-null object
production_companies    9836 non-null object
release_date            10866 non-null object
vote_count              10866 non-null int64
vote_average            10866 non-null float64
release_year            10866 non-null int64
budget_adj              1

> We will convert the budget and revenue columns to float. We will not be taking the inflation-adjusted budget and revenue columns as we want to determine whether a movie was profitable when it was released.
<br>For question 3 however we will be using the adjusted budget and revenue columns so that we can apply the same basis for comparison.

> Finally let's dig into the numerical data and statistics:

In [32]:
df.describe()

id    popularity        budget       revenue       runtime  \
count   10866.000000  10866.000000  1.086600e+04  1.086600e+04  10866.000000   
mean    66064.177434      0.646441  1.462570e+07  3.982332e+07    102.070863   
std     92130.136561      1.000185  3.091321e+07  1.170035e+08     31.381405   
min         5.000000      0.000065  0.000000e+00  0.000000e+00      0.000000   
25%     10596.250000      0.207583  0.000000e+00  0.000000e+00     90.000000   
50%     20669.000000      0.383856  0.000000e+00  0.000000e+00     99.000000   
75%     75610.000000      0.713817  1.500000e+07  2.400000e+07    111.000000   
max    417859.000000     32.985763  4.250000e+08  2.781506e+09    900.000000   

         vote_count  vote_average  release_year    budget_adj   revenue_adj  
count  10866.000000  10866.000000  10866.000000  1.086600e+04  1.086600e+04  
mean     217.389748      5.974922   2001.322658  1.755104e+07  5.136436e+07  
std      575.619058      0.935142     12.812941  3.430616e+07  1.446325e+08  
min       10.000000      1.500000   1960.000000  0.000000e+00  0.000000e+00  
25%       17.000000      5.400000   1995.000000  0.000000e+00  0.000000e+00  
50%       38.000000      6.000000   2006.000000  0.000000e+00  0.000000e+00  
75%      145.750000      6.600000   2011.000000  2.085325e+07  3.369710e+07  
max     9767.000000      9.200000   2015.000000  4.250000e+08  2.827124e+09

> We can see from the above table that there are a number of movies where the budget was not disclosed. Alternatively there could be movies with a budget but no revenue attached to it. We will have to remove all the lines with a nil budget or nil revenue to answer question 2 and question 3.

### Data Cleaning (Replace this with more specific notes!)

> We will be working off two copies, one that will be used to answer question 1 and one for the other two questions.

In [33]:
df_q1 = df.copy()
df_qn = df.copy()

In [34]:
df_q1.shape

(10866, 21)

> For question 1 we will need to drop the duplicated lines and split the genre column that has multiple genres.

In [37]:
# Drop duplicated values
df_q1.drop((df_q1[df_q1.duplicated()==True]).index,axis=0,inplace=True)
df_q1.shape

(10865, 21)

In [38]:
# Split the 'genre' column
# Let's first isolate the rows with multiple genres
df_genres = df_q1[df_q1['genres'].str.contains('\|')==True]
# We will check what is the maximum number of lines per rows that will need to be splitted
df_genres['genres'].str.count('\|').max()

4

In [39]:
# We will take note of the number of rows of df_genres 
df_genres.shape

(8514, 21)

In [40]:
# We will make four copies to split them in 5
df_genres1 = df_genres[df_genres['genres'].str.count('\|')==1].copy()
df_genres2 = df_genres[df_genres['genres'].str.count('\|')==1].copy()
df_genres3 = df_genres[df_genres['genres'].str.count('\|')==2].copy()
df_genres4 = df_genres[df_genres['genres'].str.count('\|')==3].copy()
df_genres5 = df_genres[df_genres['genres'].str.count('\|')==4].copy()

In [41]:
# We can now split each line by genre into their respective DataFrame
df_genres1['genres'] = df_genres1['genres'].apply(lambda x: x.split('|')[0])
df_genres2['genres'] = df_genres2['genres'].apply(lambda x: x.split('|')[1])
df_genres3['genres'] = df_genres3['genres'].apply(lambda x: x.split('|')[2])
df_genres4['genres'] = df_genres4['genres'].apply(lambda x: x.split('|')[3])
df_genres5['genres'] = df_genres5['genres'].apply(lambda x: x.split('|')[4])

In [42]:
# All DataFrame are consolidated onto one
new_rows = df_genres1.append([df_genres2,df_genres3,df_genres4,df_genres5])

In [43]:
# Next we need to remove all line with a '|' sign in column 'genres' from the initial DataFrame
df_q1.drop(df_genres.index,axis=0,inplace=True)
df_q1.shape

(2351, 21)

In [44]:
# Let's add back the new rows to the initial DataFrame
df_q1_clean = df_q1.append(new_rows,ignore_index=True)

In [45]:
# Finally we can remove the blank 'genres' cells
df_q1_clean.drop(df_q1_clean[df_q1_clean['genres'].isnull()==True].index,axis=0,inplace=True)

In [46]:
# The total number of rows should now be 14278
df_q1_clean.shape

(14278, 21)

> Now we can start cleaning the second DataFrame that will be used for questions 2 and 3. We will have to remove the duplicated line, convert and clean up the budget and revenue lines.

In [47]:
# First we can remove the duplcate row as we did before
df_qn.drop((df_qn[df_qn.duplicated()==True]).index,axis=0,inplace=True)
df_qn.shape

(10865, 21)

In [22]:
# Let's convert the budget and revenue columns to float
df_qn['budget'] = df_qn['budget'].astype(float)
df_qn['revenue'] = df_qn['revenue'].astype(float)
df_qn.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10866 entries, 0 to 10865
Data columns (total 21 columns):
id                      10866 non-null int64
imdb_id                 10856 non-null object
popularity              10866 non-null float64
budget                  10866 non-null float64
revenue                 10866 non-null float64
original_title          10866 non-null object
cast                    10790 non-null object
homepage                2936 non-null object
director                10822 non-null object
tagline                 8042 non-null object
keywords                9373 non-null object
overview                10862 non-null object
runtime                 10866 non-null int64
genres                  10843 non-null object
production_companies    9836 non-null object
release_date            10866 non-null object
vote_count              10866 non-null int64
vote_average            10866 non-null float64
release_year            10866 non-null int64
budget_adj           

> We must now remove all nil data from the budget and revenue columns

In [48]:
# After removing nil budget data there should be 5169 rows left
df_qn.drop((df_qn['budget']==0).index,axis=0,inplace=True)
df_qn.shape

(0, 21)

In [49]:
# Let's apply the same process for the revenue column and we can check there are 3854 rows left
df_qn.drop((df_qn[df_qn['revenue']==0]).index,axis=0,inplace=True)
df_qn.shape

(0, 21)

In [50]:
# We can now verify that there are no nil data for the budget_adj and revenue_adj columns
df_qn[df_qn['budget_adj']==0].shape,df_qn[df_qn['revenue_adj']==0].shape

((0, 21), (0, 21))